x/wasm/IBC.md

```diff
@@ -0,0 +1,128 @@
+This documents how CosmWasm contracts are expected to interact with IBC.
+
+## General Concepts
+
+**IBC Enabled** - when instantiating a contract, we detect if it supports IBC messages
+  (this is an optional extension). If so, it is considered "IBC Enabled".
+  This info should be public on all contracts. (For mock, we assume all
+  contracts are IBC enabled)
+  
+Also, please read the [IBC Docs](https://docs.cosmos.network/master/ibc/overview.html)
+for detailed descriptions of the terms *Port*, *Client*, *Connection*,
+and *Channel*
+  
+## Workflow
+
+Establishing *Clients* and *Connections* is out of the scope of this
+module and must be created by the same means as for `ibc-transfer`
+(via the cli or otherwise). `x/wasm` will bind one or more *Ports* to
+handle messages for the contracts. How *Channels* are established
+is an open question for discussion.
+
+The following actions are only available for `IBCEnabled` contracts:
+
+* Dispatch `IBCSendMsg` - this sends an IBC packet over an established
+  channel. This is returned from `handle` just like any other `CosmosMsg`
+  (For mocks, we will trigger this externally, later only
+  valid contract addresses should be able to do so).
+* Handle `IBCRecvPacket` - when another chain sends a packet destined to
+  an IBCEnabled contract, this will be routed to the proper contract
+  and call an exposed function for this purpose.
+* Handle `IBCAckPacket` - The original sender of `IBCSendMsg` will
+  get this callback eventually if the message was successfully
+  processed on the other chain
+* Handle `IBCErrorPacket` - The original sender of `IBCSendMsg` will 
+  get this callback eventually if the message failed to be
+  processed on the other chain (for any reason)
+
+For mocks, all the `IBCxxxPacket` commands are routed to some
+Golang stub handler, but containing the contract address, so we
+can perform contract-specific actions for each packet.
+
+There are open questions on how exactly the various `IBCxxPacket`
+messages are routed to a contract, as well as related to the
+channel lifecycle. (Can a contract open a channel? How do we
+allow contracts to participate in protocol version negotiation
+when establishing a channel?)
+
+## Ports and Channels
+
+There are two main proposals on how to map these concepts to
+contracts. We will describe both in detail, with arguments
+pro and contra, to select one approach.
+
+### One Port per Contract
+
+This may be the most straight-forward mapping, treating each contract 
+like a module. It does lead to very long portIDs however. Make special
+attention to both the Channel establishment (which should be compatible
+with standard ICS20 modules without changes on their part), as well
+as how contracts can properly identify their counterparty.
+
+* Upon `Instantiate`, if a contract is *IBC Enabled*, we dynamically 
+  bind a port for this contract. The port name is `wasm-<contract address>`,
+  eg. `wasm-cosmos1hmdudppzceg27qsuq707tjg8rkgj7g5hnvnw29`
```

Comment:
We should return this port either with ContractInfo or some other endpoint to make it explicit. I can imagine contract port lookups can become an issue if we leave this completely to the clients. The port naming schema   may change...?